azure/shared/client.js

```diff
@@ -0,0 +1,151 @@
+    var socket = tls.connect({ port: DD_PORT, host: DD_URL });
+    socket.on('error', err => {
+        context.log.error(err.toString());
+        socket.end();
+    });
+
+    return socket;
+}
+
+function send(socket, record) {
+    return socket.write(DD_API_KEY + ' ' + JSON.stringify(record) + '\n');
+}
+
+function handleLogs(handler, logs, context) {
+    var logsType = getLogFormat(logs);
+    switch (logsType) {
+        case STRING:
+            handler(addTagsToStringLog)(logs);
+            break;
+        case JSON_OBJECT:
+            handler(addTagsToJsonLog)(logs);
+            break;
+        case STRING_ARRAY:
+            logs.forEach(handler(addTagsToStringLog));
+            break;
+        case JSON_RECORDS:
+            logs.forEach(message => {
+                message.records.forEach(handler(addTagsToJsonLog));
+            });
+            break;
+        case JSON_ARRAY:
+            logs.forEach(handler(addTagsToJsonLog));
+            break;
+        case INVALID:
+        default:
+            context.log.warn('logs format is invalid');
+            break;
+    }
+}
+
+function getLogFormat(logs) {
+    if (typeof logs === 'string') {
+        return STRING;
+    }
+    if (!Array.isArray(logs) && typeof logs === 'object' && logs !== null) {
+        return JSON_OBJECT;
+    }
+    if (!Array.isArray(logs)) {
+        return INVALID;
+    }
+    if (typeof logs[0] === 'object') {
+        if (logs[0].records !== undefined) {
+            return JSON_RECORDS;
+        } else {
+            return JSON_ARRAY;
+        }
+    }
+    if (typeof logs[0] === 'string') {
+        return STRING_ARRAY;
+    }
+    return INVALID;
+}
+
+function addTagsToJsonLog(record, context) {
+    metadata = extractResourceId(record);
+    record['ddsource'] = metadata.source || DD_SOURCE;
+    record['ddsourcecategory'] = DD_SOURCE_CATEGORY;
+    record['service'] = DD_SERVICE;
+    record['ddtags'] = metadata.tags
+        .concat([
+            DD_TAGS,
+            'forwardername:' + context.executionContext.functionName
+        ])
+        .filter(Boolean)
+        .join(',');
+    return record;
+}
+
+function addTagsToStringLog(stringLog, context) {
+    jsonLog = { message: stringLog };
+    return addTagsToJsonLog(jsonLog, context);
+}
+
+function extractResourceId(record) {
+    metadata = { tags: [], source: '' };
+    if (
+        record.resourceId === undefined ||
+        typeof record.resourceId !== 'string' ||
+        !record.resourceId.toLowerCase().startsWith('/subscriptions/')
+    ) {
+        return metadata;
+    }
+    var resourceId = record.resourceId.toLowerCase().split('/');
+    if (resourceId.length > 2) {
+        metadata.tags.push('subscription_id:' + resourceId[2]);
+    }
+    if (resourceId.length > 4) {
+        metadata.tags.push('resource_group:' + resourceId[4]);
+    }
+    if (resourceId.length > 6) {
+        metadata.source = resourceId[6].replace('microsoft.', 'azure.');
```

Comment:
not sure if you want/need a null/undefined check on `resourceId[6]`?

e.g.
```
const x = [,,,,,,,]; 
// x.length == 7
// x[6].replace() => exception thrown
```

Reply:
A priori those are Azure provided strings and shouldn't be too weird, then the array is generated by `String.split` so I don't expect null/undefined values here